prov/mrail/src/mrail_fabric.c

```diff
@@ -73,16 +74,36 @@ static struct fi_ops_fabric mrail_fabric_ops = {
 int mrail_fabric_open(struct fi_fabric_attr *attr, struct fid_fabric **fabric,
 		      void *context)
 {
-	struct fi_info *fi;
+	struct fi_info *fi, *info, *hints;
 	struct mrail_fabric *mrail_fabric;
 	size_t i;
 	int ret;
 
 	mrail_fabric = calloc(1, sizeof(*mrail_fabric));
 	if (!mrail_fabric)
 		return -FI_ENOMEM;
+	ret = mrail_get_info_cached(attr->name, attr->prov_name, &info);
 
-	mrail_fabric->info = mrail_get_info_cached(attr->name);
+	if (ret) {
+		FI_WARN(&mrail_prov,
+			FI_LOG_CORE,
+			"No cached data found, running getinfo with prov name\n");
```

Comment:
This shouldn't happen. fi_getinfo should always be called before opening a fabric.

Reply:
This is the case for where whatever is inside the cache does not match the fabric attr given. Say two getinfo calls are given. The first getinfo is overridden (because we only cache one), but that information is given on fabric open, then instead of failing we just repopulate the cache with a new getinfo and use that instead on fabric open.

Reply:
If we do this, we lose all the hints that was given in fi_getinfo. Either we need to support only one cache entry at a time or support caching multiple entries.

Reply:
Hmm, that's true. Alright, I'll remove this for now and put a warning when overriding the cache. When we create a new address for src_addr, I'm planning on re-implementing some things (Get rid of the list format and condense the information into one info struct). Maybe we can re-add caching multiple getinfo's at that time.

Reply:
> Get rid of the list format and condense the information into one info struct

We can do this for uniform rails (ports on the same NIC).